### DISCUSSION TITLE
Guidance reborn

### DISCUSSION THREAD

**Discussion by slundberg** [guidance-ai/guidance#429] [Announcements]
You may have noticed that guidance got a new big upgrade today. `v0.1` is a dramatically new version developed while releases had to be paused over the summer. We will release a more detailed summary of the changes later but briefly:

- All guidance programs are now pure python programs. No more worrying about a distinction between "user code" in python and "template code" in handlebars.
- In addition to now being simple python functions, guidance programs are also now a superset of regular expressions and context free grammars. Allowing extremely powerful specifications to be built up incrementally.
- A new immutable model object sits at the core of all guidance programs now and manages all the state. You essentially can add any grammar to a model object and get a new model object that represents the extension of that model's state executed through that grammar.
- And a lot more!

While we have a good number of unit tests there are bound to be a few surprises, so let us know if you run into any of them.

**Comment by simon-lund** [comment]
Welcome back from the dead 🧟

**Comment by AaronWard** [comment]
Any ETA on the example notebooks being updated? appears the current ones no longer work and i want to start using guidance.

**Reply by marcotcr** [reply]
We are slowly working our way through them. Scott updated [tutorial.ipynb](https://github.com/guidance-ai/guidance/blob/main/notebooks/tutorial.ipynb), I just updated [chat_with_search.ipynb](https://github.com/guidance-ai/guidance/blob/main/notebooks/chat_with_search.ipynb).

Between those two and our new README, there should be plenty of examples to help you get started :)

**Reply by marcotcr** [reply]
Code generation example [here](https://github.com/guidance-ai/guidance/blob/main/notebooks/code_generation.ipynb)

**Reply by erlebach** [reply]
Perhaps you could post a list of the updated notebooks in README.md?

**Comment by Sam1320** [comment]
I'm so happy guidance its back!. I was already unhappily coming to terms with the idea of settling for LangChain.

**Comment by cpcdoy** [comment]
Nice! I'll have to check it out and compare to my current LMQL implementations.

**Comment by Danivilanova** [comment]
Hello @slundberg. I'm glad you are back.

With the previous version I was using a RunPod Serverless endpoint to execute a model with guidance and generate output, which includes text and variables. With handlebars, it was straightforward because I sent the entire program in text format and returned the output. How can I achieve similar efficiency using this new format?

**Reply by slundberg** [reply]
Great question that we are still sorting out. The most general serialization format is actually just python source code (or WebAssembly). If however you make purely stateless functions we could define a simple grammar syntax and ship that (this is just a potential at the moment, since we have not defined that syntax yet).

**Comment by SanderMoon** [comment]
Good lord, this looks promising. I was starting to think this lib was abandoned, so happy with this new version. 
Nice work!

**Comment by ruchira** [comment]
Congratulations! A big thank you to the team for all your work on the upgrade!

**Comment by yihanaira** [comment]
Wooo, seems like a drastic change. I have some internal demo built on 0.0.64, seems need to have some learning to make it work with the new version!

**Comment by CorentinvdBdO** [comment]
So far it has been an improvement in every level. Clearer, no pseudo code, no memory leak.

**Comment by wiiiktor** [comment]
Just adding my warm words "thank you!" :)

**Comment by wmiller** [comment]
I am so excited for this. Well done!

**Comment by moebius-ansa** [comment]
Hi.

We're using the previous version of guidance pretty extensively in our app, and I'm finding it difficult to understand how to migrate from the handlebars version to the latest one.

The previous docs for example, had an agent implementation that we're using. 

From the docs: "you can combine the await tag with geneach (which generates a list) to create an agent easily ..."

We're using both of these previous functionalities pretty extensively, but I'm having trouble understanding how to adjust existing code to achieve the same functionality with the new version.

Another question I have is whether this new version supports the latest openai API version, and will continue to do so in the future? 

It would be extremely helpful if a migration guide and a 1:1 port over of examples provided previously, especially related to geneach and await would be provided. 

This will help people using guidance to upgrade easily. Currently, I'm using a pretty old version of guidance because it does not appear to work with openai updates and the python implementation versus handlebars, and how to move from one to the other, especially around agents, isn't really explained well.

Thanks.
